### ISSUE TITLE
Photo processing / ARP Ping activity

### ISSUE THREAD

**Issue by jmai444** [FutureSharks/rpi-security#11]
Unsure if this is an issue, but there is a great deal of churning when processing photos.  There is nearly five minutes of delay before a photo is processed.  It seems to be related to the ARP ping false positive checking.  Additionally, debug messages are stating that one of my non-monitored devices is sometimes responding to the ARP ping.  192.168.1.101 is my desktop with fixed IP address - its MAC is not one I've put in the config file.  

Interesting facts: 
- The first photo to be processed is the first photo taken - I'm not sure if this means that all the others would have been processed eventually.  I shut down the service thinking no photos were going to process.
- You can see that I have added a "motion detected" message.  This is why I noticed the issue - there is a very long delay between the message and the first photo coming through.

See the syslog: 

```
Sep 25 18:25:34 raspberrypi systemd[1]: Starting rpi-security service...
Sep 25 18:25:34 raspberrypi systemd[1]: Started rpi-security service.
Sep 25 18:25:36 raspberrypi rpi-security.py:MainThread State file read: /var/lib/rpi-security/state.yaml
Sep 25 18:25:37 raspberrypi rpi-security.py:MainThread Calculated network: 192.168.1.0/24
Sep 25 18:25:54 raspberrypi kernel: [  667.995449] gpiomem-bcm2835 20200000.gpiomem: gpiomem device opened.
Sep 25 18:25:55 raspberrypi rpi-security.py:monitor_alarm_state thread running
Sep 25 18:25:55 raspberrypi rpi-security.py:capture_packets thread running
Sep 25 18:25:55 raspberrypi rpi-security.py:process_photos thread running
Sep 25 18:25:55 raspberrypi rpi-security.py:telegram_bot thread running
Sep 25 18:25:55 raspberrypi kernel: [  668.871548] device mon0 entered promiscuous mode
Sep 25 18:25:55 raspberrypi kernel: [  668.891736] device mon0 left promiscuous mode
Sep 25 18:25:55 raspberrypi kernel: [  668.923962] device mon0 entered promiscuous mode
Sep 25 18:25:57 raspberrypi rpi-security.py:MainThread rpi-security running
Sep 25 18:25:59 raspberrypi rpi-security.py:MainThread Telegram message Sent: "rpi-security: *running*"
Sep 25 18:26:25 raspberrypi kernel: [  698.775901] device wlan0 entered promiscuous mode
Sep 25 18:26:35 raspberrypi kernel: [  708.774992] device wlan0 left promiscuous mode
Sep 25 18:26:35 raspberrypi rpi-security.py:monitor_alarm_state MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:26:40 raspberrypi kernel: [  713.826129] device wlan0 entered promiscuous mode
Sep 25 18:26:51 raspberrypi kernel: [  725.170035] device wlan0 left promiscuous mode
Sep 25 18:26:51 raspberrypi rpi-security.py:monitor_alarm_state MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:26:56 raspberrypi rpi-security.py:monitor_alarm_state rpi-security is now armed
Sep 25 18:26:57 raspberrypi rpi-security.py:monitor_alarm_state Telegram message Sent: "rpi-security: *armed*"
Sep 25 18:27:04 raspberrypi rpi-security.py:Dummy-1 Motion detected
Sep 25 18:27:04 raspberrypi rpi-security.py:Dummy-1 Telegram message Sent: "rpi-security: *motion detected!*"
Sep 25 18:27:05 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182704-0.jpeg
Sep 25 18:27:06 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182704-1.jpeg
Sep 25 18:27:07 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182704-2.jpeg
Sep 25 18:27:08 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182704-3.jpeg
Sep 25 18:27:09 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182704-4.jpeg
Sep 25 18:27:10 raspberrypi kernel: [  743.876571] device wlan0 entered promiscuous mode
Sep 25 18:27:17 raspberrypi rpi-security.py:Dummy-1 Motion detected
Sep 25 18:27:18 raspberrypi rpi-security.py:Dummy-1 Telegram message Sent: "rpi-security: *motion detected!*"
Sep 25 18:27:19 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182718-0.jpeg
Sep 25 18:27:20 raspberrypi kernel: [  754.108334] device wlan0 left promiscuous mode
Sep 25 18:27:20 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f responded to ARP ping with address 192.168.1.101
Sep 25 18:27:20 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182718-1.jpeg
Sep 25 18:27:21 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182718-2.jpeg
Sep 25 18:27:22 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182718-3.jpeg
Sep 25 18:27:22 raspberrypi kernel: [  756.190181] device wlan0 entered promiscuous mode
Sep 25 18:27:23 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182718-4.jpeg
Sep 25 18:27:24 raspberrypi rpi-security.py:Dummy-1 Motion detected
Sep 25 18:27:25 raspberrypi rpi-security.py:Dummy-1 Telegram message Sent: "rpi-security: *motion detected!*"
Sep 25 18:27:26 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182725-0.jpeg
Sep 25 18:27:27 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182725-1.jpeg
Sep 25 18:27:29 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182725-2.jpeg
Sep 25 18:27:30 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182725-3.jpeg
Sep 25 18:27:31 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-182725-4.jpeg
Sep 25 18:27:32 raspberrypi kernel: [  766.173209] device wlan0 left promiscuous mode
Sep 25 18:27:32 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:27:34 raspberrypi kernel: [  768.224440] device wlan0 entered promiscuous mode
Sep 25 18:27:44 raspberrypi kernel: [  777.667628] device wlan0 left promiscuous mode
Sep 25 18:27:44 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:27:44 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:27:49 raspberrypi kernel: [  782.725707] device wlan0 entered promiscuous mode
Sep 25 18:27:58 raspberrypi kernel: [  792.248501] device wlan0 left promiscuous mode
Sep 25 18:27:58 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:28:00 raspberrypi kernel: [  794.306233] device wlan0 entered promiscuous mode
Sep 25 18:28:09 raspberrypi kernel: [  803.006495] device wlan0 left promiscuous mode
Sep 25 18:28:09 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:28:11 raspberrypi kernel: [  805.078611] device wlan0 entered promiscuous mode
Sep 25 18:28:20 raspberrypi kernel: [  814.000448] device wlan0 left promiscuous mode
Sep 25 18:28:20 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f responded to ARP ping with address 192.168.1.101
Sep 25 18:28:20 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:28:25 raspberrypi kernel: [  819.059234] device wlan0 entered promiscuous mode
Sep 25 18:28:34 raspberrypi kernel: [  828.145391] device wlan0 left promiscuous mode
Sep 25 18:28:34 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:28:36 raspberrypi kernel: [  830.197810] device wlan0 entered promiscuous mode
Sep 25 18:28:45 raspberrypi kernel: [  839.299724] device wlan0 left promiscuous mode
Sep 25 18:28:45 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:28:47 raspberrypi kernel: [  841.348005] device wlan0 entered promiscuous mode
Sep 25 18:28:56 raspberrypi kernel: [  849.803958] device wlan0 left promiscuous mode
Sep 25 18:28:56 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:28:56 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:29:01 raspberrypi kernel: [  854.864657] device wlan0 entered promiscuous mode
Sep 25 18:29:10 raspberrypi kernel: [  863.721609] device wlan0 left promiscuous mode
Sep 25 18:29:10 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f responded to ARP ping with address 192.168.1.101
Sep 25 18:29:12 raspberrypi kernel: [  865.770097] device wlan0 entered promiscuous mode
Sep 25 18:29:22 raspberrypi kernel: [  875.745668] device wlan0 left promiscuous mode
Sep 25 18:29:22 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f responded to ARP ping with address 192.168.1.101
Sep 25 18:29:24 raspberrypi kernel: [  877.799019] device wlan0 entered promiscuous mode
Sep 25 18:29:32 raspberrypi kernel: [  886.462998] device wlan0 left promiscuous mode
Sep 25 18:29:32 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:29:32 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:29:37 raspberrypi kernel: [  891.526744] device wlan0 entered promiscuous mode
Sep 25 18:29:46 raspberrypi kernel: [  899.852828] device wlan0 left promiscuous mode
Sep 25 18:29:46 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:29:48 raspberrypi kernel: [  901.900666] device wlan0 entered promiscuous mode
Sep 25 18:29:48 raspberrypi systemd[1]: Starting Cleanup of Temporary Directories...
Sep 25 18:29:48 raspberrypi systemd[1]: Started Cleanup of Temporary Directories.
Sep 25 18:29:58 raspberrypi kernel: [  912.307222] device wlan0 left promiscuous mode
Sep 25 18:29:58 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:30:00 raspberrypi kernel: [  914.360230] device wlan0 entered promiscuous mode
Sep 25 18:30:10 raspberrypi kernel: [  924.276652] device wlan0 left promiscuous mode
Sep 25 18:30:10 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f responded to ARP ping with address 192.168.1.101
Sep 25 18:30:10 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:30:15 raspberrypi kernel: [  929.327862] device wlan0 entered promiscuous mode
Sep 25 18:30:24 raspberrypi kernel: [  938.284829] device wlan0 left promiscuous mode
Sep 25 18:30:24 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f responded to ARP ping with address 192.168.1.101
Sep 25 18:30:26 raspberrypi kernel: [  940.334195] device wlan0 entered promiscuous mode
Sep 25 18:30:35 raspberrypi kernel: [  949.026329] device wlan0 left promiscuous mode
Sep 25 18:30:35 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:30:37 raspberrypi kernel: [  951.075062] device wlan0 entered promiscuous mode
Sep 25 18:30:50 raspberrypi kernel: [  963.802832] device wlan0 left promiscuous mode
Sep 25 18:30:50 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:30:50 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:30:55 raspberrypi kernel: [  968.857376] device wlan0 entered promiscuous mode
Sep 25 18:31:09 raspberrypi kernel: [  982.790253] device wlan0 left promiscuous mode
Sep 25 18:31:09 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:31:11 raspberrypi kernel: [  984.886270] device wlan0 entered promiscuous mode
Sep 25 18:31:24 raspberrypi kernel: [  997.716863] device wlan0 left promiscuous mode
Sep 25 18:31:24 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f responded to ARP ping with address 192.168.1.101
Sep 25 18:31:26 raspberrypi kernel: [  999.773546] device wlan0 entered promiscuous mode
Sep 25 18:31:36 raspberrypi kernel: [ 1010.000116] device wlan0 left promiscuous mode
Sep 25 18:31:36 raspberrypi rpi-security.py:process_photos MAC aa:aa:aa:aa:aa:8f did not respond to ARP ping
Sep 25 18:31:36 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:31:36 raspberrypi rpi-security.py:process_photos Processing the photo: /var/tmp/rpi-security-2016-09-25-182704-0.jpeg
Sep 25 18:31:57 raspberrypi systemd[1]: Stopping rpi-security service...
Sep 25 18:32:02 raspberrypi rpi-security.py:MainThread rpi-security stopping...
Sep 25 18:32:02 raspberrypi rpi-security.py:MainThread rpi-security stopping...
Sep 25 18:32:04 raspberrypi kernel: [ 1037.849649] device mon0 left promiscuous mode
Sep 25 18:32:04 raspberrypi avahi-daemon[334]: Withdrawing workstation service for mon0.
Sep 25 18:32:04 raspberrypi systemd[1]: Stopped rpi-security service.
```

I am looking at this code in process_photos:

```
else:
                logger.debug('Starting to process photos')
                for photo in list(captured_photos):
                    if time.time() - alarm_state['last_packet'] < 300:
                        break
                    logger.debug('Processing the photo: %s' % photo)
```

If I am reading this right, it is doing nothing until 5 minutes after the last packet was detected?

**Comment by jmai444** [comment]
I have commented out the if / break statement from the code shown above and it does indeed start processing photos almost immediately.

Still not sure about the false ARP ping response messages.  I haven't noticed that they cause incorrect operation, but they appear fairly often.

**Comment by FutureSharks** [comment]
Hey jmai444,
Thanks for the feedback!

The `if / break` statement simply makes it stop processing photos if a packet is detected. Sending photos can take a while, if there is a false trigger and then a packet is detected after a delay and after processing photos has already started then this just stops sending the remaining photos. Removing this statement won't make the photos processed quicker and it does not wait 5 mins. If you want quick photo processing (at the risk of false positives) then comment this line:
https://github.com/FutureSharks/rpi-security/blob/master/bin/rpi-security.py#L189

By design, it should behave like this....

_These will always apply_
a. When transitioning to armed state, try to ARP ping macs (stops false arming).
b. When motion is detected and state is armed, take 5 photos. 

_The process_
1. If there are new captured photos then ARP ping our MACs 3 times (stops false alarms when returning home). If we get replies then discard all photos and disarm.
2. If step 1 doesn't detect the MACs then send each photo via telegram. If at any point our packets from our MACs are detected when processing photos then just stop processing photos and discard any that are unsent.

Here is a log from mine with some comments:

```
Sep 25 18:44:22 raspberrypi rpi-security.py:MainThread Telegram message Sent: "rpi-security running"
(now transitioning to armed)
Sep 25 18:56:09 raspberrypi rpi-security.py:monitor_alarm_state MAC 70:14:a6:e2:14:7a did not respond to ARP ping
Sep 25 18:56:23 raspberrypi rpi-security.py:Dummy-1 Motion detected but current_state is: disarmed
Sep 25 18:56:28 raspberrypi rpi-security.py:monitor_alarm_state MAC 70:14:a6:e2:14:7a did not respond to ARP ping
Sep 25 18:56:33 raspberrypi rpi-security.py:monitor_alarm_state rpi-security is now armed
Sep 25 18:56:33 raspberrypi rpi-security.py:monitor_alarm_state Telegram message Sent: "rpi-security: *armed*"
Sep 25 18:57:16 raspberrypi rpi-security.py:Dummy-1 Motion detected
Sep 25 18:57:17 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-185716-0.jpeg
Sep 25 18:57:18 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-185716-1.jpeg
Sep 25 18:57:19 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-185716-2.jpeg
Sep 25 18:57:20 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-185716-3.jpeg
Sep 25 18:57:21 raspberrypi rpi-security.py:Dummy-1 Captured image: /var/tmp/rpi-security-2016-09-25-185716-4.jpeg
Sep 25 18:57:36 raspberrypi rpi-security.py:process_photos MAC 70:14:a6:e2:14:7a did not respond to ARP ping
Sep 25 18:57:51 raspberrypi rpi-security.py:process_photos MAC 70:14:a6:e2:14:7a did not respond to ARP ping
Sep 25 18:58:07 raspberrypi rpi-security.py:process_photos MAC 70:14:a6:e2:14:7a did not respond to ARP ping
(MAC did not reply so starting to send photos)
Sep 25 18:58:07 raspberrypi rpi-security.py:process_photos Starting to process photos
Sep 25 18:58:07 raspberrypi rpi-security.py:process_photos Processing the photo: /var/tmp/rpi-security-2016-09-25-185716-0.jpeg
Sep 25 18:58:18 raspberrypi rpi-security.py:process_photos Telegram file sent: /var/tmp/rpi-security-2016-09-25-185716-0.jpeg
Sep 25 18:58:18 raspberrypi rpi-security.py:process_photos Archiving of photo complete: /var/tmp/rpi-security-2016-09-25-185716-0.jpeg
Sep 25 18:58:19 raspberrypi rpi-security.py:process_photos Processing the photo: /var/tmp/rpi-security-2016-09-25-185716-1.jpeg
(I turned on my wifi at this point)
Sep 25 18:58:26 raspberrypi rpi-security.py:capture_packets Packet detected from 70:14:a6:e2:14:7a
Sep 25 18:58:27 raspberrypi rpi-security.py:process_photos Telegram file sent: /var/tmp/rpi-security-2016-09-25-185716-1.jpeg
Sep 25 18:58:27 raspberrypi rpi-security.py:process_photos Archiving of photo complete: /var/tmp/rpi-security-2016-09-25-185716-1.jpeg
(Moving to disarmed since a packet was detected)
Sep 25 18:58:28 raspberrypi rpi-security.py:monitor_alarm_state rpi-security is now disarmed
Sep 25 18:58:28 raspberrypi rpi-security.py:monitor_alarm_state Telegram message Sent: "rpi-security: *disarmed*"
(removing all unprocessed photos as the system is now disarmed)
Sep 25 18:58:32 raspberrypi rpi-security.py:process_photos Removing photo as it is a false positive: /var/tmp/rpi-security-2016-09-25-185716-2.jpeg
```

You can see that motion was detected at 18:57:16 and the first message with a photo was sent at 18:58:18. That's 1 minute. It will be longer if you have more MACs in your config but it shouldn't be 5 mins and it looks like you also have a single MAC.

The reason for the many extra ARP pings is that I found both the iPhone and Android phones are quite unreliable and don't always reply to ARP pings or send frequent wifi scan packets (that we detect). Sometimes when a phone is not in use it stops sending the packets we are detecting for 15-30 mins. And then occasionally they won't reply to ARP ping reliably even though they are on the network. When you combine both of these behaviours I had quite a lot of false positives. I found that even a 1-2% false positive rate was really annoying.

> non-monitored devices is sometimes responding to the ARP ping. 192.168.1.101 is my desktop with fixed IP address

I get this too on occasion too, it's my Windows 10 desktop. I don't really know why though. Perhaps this is causing the issue and why it seems to try and ARP ping your MAC for 5 whole minutes. I'll have think...

**Comment by FutureSharks** [comment]
What OS is that desktop?

Could you revert rpi-security.py changes, disconnect the desktop, and try again? Then send a new log, hopefully without these lines:
`MAC 44:74:6c:48:54:8f responded to ARP ping with address 192.168.1.101`

If that works, I think I could update arp_ping_macs to filter replies from other hosts.

**Comment by jmai444** [comment]
Thanks for the replies.

Understanding your description, I see the issue.  The five minute wait I experience is due to the temporary test parameters I use in the config file.  I set the time to arm = 30 seconds after last detecting a packet so that I did not need to wait the full 12 minutes in order to test the motion detected behaviour.  This causes the repeated arp pings because the if / break kicks in.  It doesn't flag any photos as false positives, so it still processes them all after the 5 minute delay.  It is not an issue when using normal parameters in the config file.

The other issue with the spurious message involving 192.168.1.101 - It's a Linux box, but I am not sure it's always the one identified in the message.  I think I have seen other devices at other times.  It can't be detecting the actual MAC in the config file.  The device was switched off.  

It doesn't seem to be an issue interfering with operation.  I noticed the spurious message immediately before the "rpi-security is armed" message.   Perhaps a bug in the module returning the scan data?

When I have a chance I will revert my changes out and see if I can duplicate the issue.

**Comment by FutureSharks** [comment]
Ah now I understand. Setting time to arm to 30 seconds won't work because there are other parts of the code assuming it's much longer than 30 seconds.

e.g.
https://github.com/FutureSharks/rpi-security/blob/master/bin/rpi-security.py#L190
https://github.com/FutureSharks/rpi-security/blob/master/bin/rpi-security.py#L198

I should probably make that clearer or set a minimum value of 300. With wifi being sometimes unreliable, mobile phones also being slightly unreliable and sometimes images can take a min or two to finish sending, I've had to make a lot of room for these inconsistencies. This shows in longer timeouts and many ARP pings to be double sure before sending alerts. You could try 300 but I have a feeling you might get false positives.

**Comment by jmai444** [comment]
Changing to 30 seconds was merely for testing.  I will just need to be more patient while testing!

Maybe it is my network set up, but after more than one week's continuous use the system has never armed itself while we were home - not even overnight.  I might be able to get away without checking for false positives during the photo processing.

**Comment by FutureSharks** [comment]
OK cool. Let me know how it goes.

**Comment by jmai444** [comment]
When I run tests in python interpreter I sometimes get an answer to the srp() statement when pinging a non-existent MAC address - when this happens, the returned ip address is the one for my Linux desktop which is not one I have included in the config file to monitor. This occurs once in three or four pings:

```
>>> answered,unanswered = srp(Ether(dst='aa:aa:aa:aa:aa:96')/ARP(pdst='192.168.1.0/24'), timeout=1, verbose=True)
Begin emission:
Finished to send 256 packets.

Received 41 packets, got 1 answers, remaining 255 packets
>>> for reply in answered:
...      result = []
...      result.append(str(reply[0].pdst))
...      result = ', '.join(result)
... 
>>> print result
192.168.1.101
```

**Comment by FutureSharks** [comment]
Interesting that you have this for a linux host and I had it for a Windows Host. Perhaps it's a scapy issue.

If you disconnect the 192.168.1.101 host from the Wifi, do you still get this rogue reply?

**Comment by FutureSharks** [comment]
Can you try this a few times?

```
>>> answered,unanswered = srp(Ether(dst='c4:3a:be:1f:bc:96')/ARP(pdst='192.168.1.0/24'), timeout=1, verbose=True)
>>> answered[0][1].hwsrc
(What MAC here?)
```

Do you get the MAC of the Linux host sometimes?

**Comment by jmai444** [comment]
I was misremembering which machine that IP address belongs to.  The IP address and the MAC address shown by running your latest belong to my file server.  It is running an OpenSolaris variant called OpenIndiana.  The file server is hosting Samba shares as well as NFS shares.  Perhaps Samba is the culprit?

The bash utility arp-scan never gets a reply if the MAC address doesn't exist.  I tried a few dozen times.  Maybe scapy is reporting additional or spurious replies?

```
xxxxxx@xxxxx:~$ sudo arp-scan --destaddr=aa:aa:aa:aa:aa:96 --interface=eth0 192.168.1.0/24
Interface: eth0, datalink type: EN10MB (Ethernet)
Starting arp-scan 1.8.1 with 256 hosts (http://www.nta-monitor.com/tools/arp-scan/)

0 packets received by filter, 0 packets dropped by kernel
Ending arp-scan 1.8.1: 256 hosts scanned in 1.327 seconds (192.92 hosts/sec). 0 responded
```

**Comment by jmai444** [comment]
If I introduce a delay between packets using the inter parameter with srp(), scapy nearly always shows a reply from my file server.

It takes considerably longer to do the scan, but I reckon this is the reason you sometimes don't get responses to the ping from your monitored devices.  scapy doesn't handle them correctly without the delay between packets.  As I increase the delay, scapy reports more packets received.

Still doesn't resolve the mystery of why my file server responds!

```
>>> answered,unanswered = srp(Ether(dst='aa:aa:aa:aa:aa:96')/ARP(pdst='192.168.1.0/24'), timeout=1, verbose=True, inter=0.1)
Begin emission:
Finished to send 256 packets.

Received 128 packets, got 1 answers, remaining 255 packets
>>> answered[0][1].hwsrc
'aa:aa:aa:aa:aa:41'
```

If I scan for a MAC address that exists on the network, I usually get two answers one being the file server, the other being the monitored device.

```
>>> answered,unanswered = srp(Ether(dst='aa:aa:aa:aa:aa:97')/ARP(pdst='192.168.1.0/24'), timeout=1, verbose=True, inter=0.1)
Begin emission:
Finished to send 256 packets.

Received 64 packets, got 2 answers, remaining 254 packets
>>> answered[1][1].hwsrc
'aa:aa:aa:aa:aa:97'
>>> answered[0][1].hwsrc
'aa:aa:aa:aa:aa:41'
```

**Comment by jmai444** [comment]
I have an idea I might try.  Another thread that mostly sleeps, but once no packets are received for, say, five minutes, it wakes up and performs the mac_arp_ping and continues to do so, say once per minute until a packet is received by capture_packets.  Then the other threads wont need to stop what they are doing to perform the mac_arp_ping.

**Comment by jmai444** [comment]
Sorry for the information overload, but another piece of data:

srp() as issued in your code (perhaps always?) puts wlan0 in promiscuous mode. Issuing arp-scan from the command line does not from messages I see in /var/log/syslog.

**Comment by jmai444** [comment]
OK, I imported scapy conf and set promisc=0 and sniff_promisc=0.  wlan0 does not enter promiscuous mode to scan, but it made no difference to the results.  I even got three responses once - my router responded.  

Interestingly, I have a samba shared disk on the router USB port as well.  It might also have something to do with the Windows Computer Browser Service. See link:
[http://www.rhyshaden.com/nt_prot.htm](url)

**Comment by FutureSharks** [comment]
OK  quite a lot going on here so I might break it down a bit.
1. **Your original issue**
   I think this is resolved? You set `packet_timeout` too low and now I think the issue is sorted.
2. **False replies when using scapy.srp**
   Although I've had this in the past I cannot reproduce it at all anymore, even with `inter=0.1`. I've pushed a change to stop the rogue replies (or maybe it's a scapy bug) from affecting the _arp_ping function: https://github.com/FutureSharks/rpi-security/pull/12
   I just testing it now.
3. **Setting promisc=0 and sniff_promisc=0**
   Yes I guess we can set promisc=0 and sniff_promisc=0 since it's not really required and still stop annoying log messages. I'm not sure how to set it without importing the whole module though?

**Comment by jmai444** [comment]
1. Correct.
2. Thanks.
3. Not sure what you mean about importing the whole module. I imported conf in addition to srp, ARP, Ether, etc. during an interactive session.  This allowed me to set the variables.  I am not sure how to stop it from also impacting sniff() which needs to be in promiscuous mode.  Is that what you mean?

**Comment by FutureSharks** [comment]
1. I've merged that PR now.
2. It doesn't work for me.

```
>>> from scapy.all import config, srp, Ether, ARP
>>> config.promisc=0
>>> config.sniff_promisc=0
>>> answered,unanswered = srp(Ether(dst='dd:3a:be:1f:bc:97')/ARP(pdst='192.168.1.0/24'), timeout=1, verbose=True)
Begin emission:
Finished to send 256 packets.
Received 11 packets, got 0 answers, remaining 256 packets
>>>
```

I still get this in the logs:

```
Sep 28 19:25:31 raspberrypi kernel: [269973.658196] device wlan0 entered promiscuous mode
Sep 28 19:25:45 raspberrypi kernel: [269987.921436] device wlan0 left promiscuous mode
```

Does it work for you?

**Comment by jmai444** [comment]
I imported conf (not config).  I get an error trying to import config:

```
>>> from scapy.all import config, srp, Ether, ARP
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: cannot import name config
```

This works for me, no messages in the log:

```
>>> from scapy.all import srp, Ether, ARP, conf
>>> conf.promisc=0
>>> conf.sniff_promisc=0
>>> answered,unanswered = srp(Ether(dst='dd:3a:be:1f:bc:97')/ARP(pdst='192.168.1.0/24'), timeout=1, verbose=True)
Begin emission:
Finished to send 256 packets.

Received 37 packets, got 1 answers, remaining 255 packets
```

**Comment by FutureSharks** [comment]
Strange how you get an error and I don't when importing config. Do you have scapy version 2.3.2.dev0?

I've made a PR to disable promiscuous setting, now testing it:
https://github.com/FutureSharks/rpi-security/pull/13

**Comment by jmai444** [comment]
I'll have to check version of scapy when I get home, but I followed your instructions to get the latest when setting everything up two weeks ago.  Should have nearly the latest.

Thanks for your efforts on this.  It really does work nicely.

One thing I have changed is the captured photo size.  I reduced resolution to 1280 x 720 (you lose some top and bottom of the frame, but that is not important where I have the camera set up.)  And when I capture I set quality=15.  This reduces the size to ~100Kb.  Makes the photo capture and subsequent handling much snappier.
